tests/headless/test_cdn.py

```diff
@@ -0,0 +1,470 @@
+
+
+def is_cdn_cache_miss(response):
+    """Checks the response for evidence of a cache miss on the CDN."""
+    return is_cloudfront_cache_miss(response)
+
+
+def is_cdn_error(response):
+    """Checks the response for evidence of an error from the CDN."""
+    return is_cloudfront_error(response)
+
+
+def request(method, url, **kwargs):
+    if 'timeout' not in kwargs:
+        kwargs.update(timeout=DEFAULT_TIMEOUT)
+    if 'allow_redirects' not in kwargs:
+        kwargs.update(allow_redirects=False)
+    return requests.request(method, url, **kwargs)
+
+
+def assert_never_cached(url, expected_status_code=200, method='get',
+                        check_cache_header=True, **request_kwargs):
+    response = request(method, url, **request_kwargs)
+    assert response.status_code == expected_status_code
+    if check_cache_header:
+        assert 'no-cache' in response.headers['Cache-Control']
+        assert 'no-store' in response.headers['Cache-Control']
+        assert 'must-revalidate' in response.headers['Cache-Control']
+        assert 'max-age=0' in response.headers['Cache-Control']
+    if expected_status_code >= 400:
+        assert is_cdn_error(response)
+    else:
+        assert is_cdn_cache_miss(response)
+    return response
+
+
+def assert_not_cached_by_cdn(url, expected_status_code=200, method='get',
+                             check_cache_header=True, **request_kwargs):
+    response1 = assert_never_cached(url, expected_status_code, method=method,
+                                    check_cache_header=check_cache_header,
+                                    **request_kwargs)
+    response2 = assert_never_cached(url, expected_status_code, method=method,
+                                    check_cache_header=check_cache_header,
+                                    **request_kwargs)
+    if expected_status_code in (301, 302):
+        assert response2.headers['location'] == response1.headers['location']
+    return response2
+
+
+def assert_cached_by_cdn(url, expected_status_code=200, method='get',
+                         **request_kwargs):
+    response = request(method, url, **request_kwargs)
+    assert response.status_code == expected_status_code
+    assert 'public' in response.headers['Cache-Control']
+    assert (('max-age' in response.headers['Cache-Control']) or
+            ('s-maxage' in response.headers['Cache-Control']))
+    if is_cdn_cache_miss(response):
+        response2 = request(method, url, **request_kwargs)
+        assert response2.status_code == expected_status_code
+        assert is_cdn_cache_hit(response2)
+        if expected_status_code == 200:
+            assert response2.content == response.content
+        elif expected_status_code in (301, 302):
+            assert (response2.headers['location'] ==
+                    response.headers['location'])
+    else:
+        assert is_cdn_cache_hit(response)
+    return response
+
+
+@pytest.mark.headless
```

Comment:
No, I don't use them. I was like a sheep following along with the other headless tests! 🐑 In fact, now that I think of it, I think we should dump them, since I don't think they're really that useful. What do you think?

Reply:
Let's remove them until we use them. I thought ``nondestructive`` was used somewhere (Jenkins?), but I think it may be the same a login tests and unused.